### PR TITLE
uDraw GameTablet device emulation

### PR DESCRIPTION
**Pull request by Florin9doi**
- uDraw GameTablet device emulation
- Requires the following Settings:
  - Common:
    - I/O -> Mouse Handler = Basic
    - I/O -> GameTablet = Enabled

  - uDraw Studio: Instant Artist
    - Pads -> Player1 = connected, Player2 = Null
  - Marvel Super Hero Squad: Comic Combat
    - Pads -> Player1 = Null, Player2 = connected
    - Disable Vertex Cache : https://forums.rpcs3.net/showthread.php?tid=207254&pid=325836#pid325836
  - uDraw Pictionary
    - any

### CODE REVIEW

File: rpcs3/Emu/Io/GameTablet.h [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+#include "Emu/Io/usb_device.h"
+
+class usb_device_gametablet : public usb_device_emulated
+{
+	u8 buttons;
+	s32 x_pos;
+	s32 y_pos;
+	s32 x_max;
+	s32 y_max;
```
Comment (Megamouse) [inline rpcs3/Emu/Io/GameTablet.h:11]:
```suggestion
	u8 buttons = 0;
	s32 x_pos = 0;
	s32 y_pos = 0;
	s32 x_max = 0;
	s32 y_max = 0;
```

File: rpcs3/Emu/Io/GameTablet.cpp [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+
+void usb_device_gametablet::interrupt_transfer(u32 buf_size, u8* buf, u32 /*endpoint*/, UsbTransfer* transfer)
+{
+	transfer->fake            = true;
+	transfer->expected_count  = 27;
+	transfer->expected_result = HC_CC_NOERR;
+	// Interrupt transfers are slow (6ms, TODO accurate measurement)
+	transfer->expected_time = get_timestamp() + 6000;
+
+	memset(buf, 0, buf_size);
```
Comment (Megamouse) [inline rpcs3/Emu/Io/GameTablet.cpp:114]:
```suggestion
	std::memset(buf, 0, buf_size);
```

File: rpcs3/Emu/Io/GameTablet.cpp [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+	buf[0x17] = 0xea; // accel Z
+	buf[0x18] = 0x01;
+
+	buf[0x1a] = 0x02;
+
+	if (!is_input_allowed())
+	{
+		return;
+	}
+	if (g_cfg.io.mouse != mouse_handler::basic)
```
Comment (Megamouse) [inline rpcs3/Emu/Io/GameTablet.cpp:142]: missing newline

File: rpcs3/Emu/Io/GameTablet.cpp [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+		gametablet_log.warning("GameTablet requires mouse_handler configured to basic");
+		return;
+	}
+
+	std::lock_guard lock(pad::g_pad_mutex);
+	const auto gamepad_handler = pad::get_current_handler();
+	const auto& pads = gamepad_handler->GetPads();
+
+	bool up, right, down, left;
+	up = right = down = left = false;
```
Comment (Megamouse) [inline rpcs3/Emu/Io/GameTablet.cpp:153]: you can do this in one line

File: rpcs3/Emu/Io/GameTablet.cpp [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+	transfer->expected_time   = get_timestamp() + 100;
+
+	// Control transfers are nearly instant
+	switch (bmRequestType)
+	{
+	case LIBUSB_ENDPOINT_OUT | LIBUSB_REQUEST_TYPE_CLASS | LIBUSB_RECIPIENT_INTERFACE: // 0x21
+		switch (bRequest)
+		{
+		case 0x09: // SET_REPORT
+			gametablet_log.trace("Leds: %s/%s/%s/%s",
```
Comment (Megamouse) [inline rpcs3/Emu/Io/GameTablet.cpp:87]:
```suggestion
			ensure(buf_size > 2);
			gametablet_log.trace("Leds: %s/%s/%s/%s",
```

File: rpcs3/Emu/Io/GameTablet.cpp [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+		usb_device_emulated::control_transfer(bmRequestType, bRequest, wValue, wIndex, wLength, buf_size, buf, transfer);
+		break;
+	}
+}
+
+extern bool is_input_allowed();
+
+void usb_device_gametablet::interrupt_transfer(u32 buf_size, u8* buf, u32 /*endpoint*/, UsbTransfer* transfer)
+{
+	transfer->fake            = true;
```
Comment (Megamouse) [inline rpcs3/Emu/Io/GameTablet.cpp:108]:
```suggestion
	ensure(buf_size > 0x1a);

	transfer->fake            = true;
```

File: rpcs3/Emu/Io/MouseHandler.h [hunk 7]
```diff
@@ -80,6 +80,10 @@ struct MouseData
 	s8 y_axis = 0;
 	s8 wheel = 0;
 	s8 tilt = 0; // (TODO)
+	s32 x_pos;
```
Comment (Megamouse) [inline rpcs3/Emu/Io/MouseHandler.h:83]: The changes in this file make no sense.
This is cellMouse related code and is not supposed to be used in this way.

File: rpcs3/Emu/Io/GameTablet.cpp [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+		buf[0x02] = 0x07;
+
+
+	auto& mouse_handler = g_fxo->get<MouseHandlerBase>();
+	std::lock_guard mouse_lock(mouse_handler.mutex);
+
+	mouse_handler.Init(1);
+
+	const MouseInfo& current_info = mouse_handler.GetInfo();
+	MouseDataList& data_list = mouse_handler.GetDataList(0);
```
Comment (Megamouse) [inline rpcs3/Emu/Io/GameTablet.cpp:244]: This makes no sense.
Just use the mouse and its members
```suggestion
	const Mouse& mouse = ::at32(handler.GetMice(), 0);
```

File: rpcs3/Emu/Io/GameTablet.cpp [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+		return;
+	}
+
+	static uint8_t noise_x = 0; // Toggle the LSB to simulate a noisy signal, Instant Artist dislikes a pen held perfectly still
+	static uint8_t noise_y = 0;
+	const int tablet_max_x = 1920;
+	const int tablet_max_y = 1080;
+
+	long tablet_x_pos = (x_pos * tablet_max_x / x_max) ^ noise_x;
+	long tablet_y_pos = (y_pos * tablet_max_y / y_max) ^ noise_y;
```
Comment (Megamouse) [inline rpcs3/Emu/Io/GameTablet.cpp:271]:
```suggestion
	const long tablet_x_pos = (x_pos * tablet_max_x / x_max) ^ noise_x;
	const long tablet_y_pos = (y_pos * tablet_max_y / y_max) ^ noise_y;
```

File: rpcs3/Emu/Io/GameTablet.cpp [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+{
+	transfer->fake            = true;
+	transfer->expected_count  = buf_size;
+	transfer->expected_result = HC_CC_NOERR;
+	transfer->expected_time   = get_timestamp() + 100;
+
+	// Control transfers are nearly instant
+	switch (bmRequestType)
+	{
+	case LIBUSB_ENDPOINT_OUT | LIBUSB_REQUEST_TYPE_CLASS | LIBUSB_RECIPIENT_INTERFACE: // 0x21
```
Comment (Megamouse) [inline rpcs3/Emu/Io/GameTablet.cpp:83]:
```suggestion
	case 0U /*silences warning*/ | LIBUSB_ENDPOINT_OUT | LIBUSB_REQUEST_TYPE_CLASS | LIBUSB_RECIPIENT_INTERFACE: // 0x21
```

File: rpcs3/Emu/Io/GameTablet.cpp (hunk 8 above)
Comment (Florin9doi) [reply rpcs3/Emu/Io/GameTablet.cpp]: done

**Review by Megamouse — COMMENTED** [review]
You have to add the new cpp files to the corresponding Cmakelists file

File: rpcs3/Emu/Io/GameTablet.cpp [hunk 11]
```diff
@@ -149,94 +149,95 @@ void usb_device_gametablet::interrupt_transfer(u32 buf_size, u8* buf, u32 /*endp
...
-	else if (down && !left && !right)
-		buf[0x02] = 0x04;
-	else if (down && left)
-		buf[0x02] = 0x05;
-	else if (left && !up && !down)
-		buf[0x02] = 0x06;
-	else if (up && left)
-		buf[0x02] = 0x07;
 
+		if (!up && !right && !down && !left)
```
Comment (Megamouse) [inline rpcs3/Emu/Io/GameTablet.cpp:222]: This part doesn't have to be in the lock scope

File: rpcs3/rpcs3qt/tooltips.h [hunk 12]
```diff
@@ -231,6 +231,7 @@ class Tooltips : public QObject
 		const QString buzz              = tr("Buzz! support.\nSelect 1 or 2 controllers if the game requires Buzz! controllers and you don't have real controllers.\nSelect Null if the game has support for DualShock or if you have real Buzz! controllers.");
 		const QString turntable         = tr("DJ Hero Turntable controller support.\nSelect 1 or 2 controllers if the game requires DJ Hero Turntable controllers and you don't have real turntable controllers.\nSelect Null if the game has support for DualShock or if you have real turntable controllers.\nA real turntable controller can be used at the same time as an emulated turntable controller.");
 		const QString ghltar            = tr("Guitar Hero Live (GHL) Guitar controller support.\nSelect 1 or 2 controllers if the game requires GHL Guitar controllers and you don't have real guitar controllers.\nSelect Null if the game has support for DualShock or if you have real guitar controllers.\nA real guitar controller can be used at the same time as an emulated guitar controller.");
+		const QString gametablet        = tr("GameTablet emulated controller support.\nDisable to use a physical GameTablet device");
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/tooltips.h:234]: You should mention that the second controller is needed

File: rpcs3/rpcs3qt/tooltips.h [hunk 13]
```diff
@@ -231,6 +231,7 @@ class Tooltips : public QObject
...
 		const QString turntable         = tr("DJ Hero Turntable controller support.\nSelect 1 or 2 controllers if the game requires DJ Hero Turntable controllers and you don't have real turntable controllers.\nSelect Null if the game has support for DualShock or if you have real turntable controllers.\nA real turntable controller can be used at the same time as an emulated turntable controller.");
 		const QString ghltar            = tr("Guitar Hero Live (GHL) Guitar controller support.\nSelect 1 or 2 controllers if the game requires GHL Guitar controllers and you don't have real guitar controllers.\nSelect Null if the game has support for DualShock or if you have real guitar controllers.\nA real guitar controller can be used at the same time as an emulated guitar controller.");
+		const QString gametablet        = tr("uDraw GameTablet emulated controller support.\nDisable to use a physical uDraw GameTablet device.\n• uDraw Studio: Instant Artist  requires to configure Mouse Handler=Basic, Pad1=Connected and Pad2=Null.\n• Marvel Super Hero Squad: Comic Combat  requires to configure Mouse Handler=Basic, Pad1=Null and Pad2=Connected.");
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/tooltips.h:234]: Why are there 2 spaces after each of the game names?
Also, I don't think it makes sense to mention specific games. There are more than 2 games that use udraw, and this is something that belongs to the wiki.
I had something more general in mind. Like a hint that you might need one of the two settings.
```suggestion
		const QString gametablet        = tr("uDraw GameTablet emulated controller support.\nDisable to use a physical uDraw GameTablet device.\n• uDraw Studio: Instant Artist  requires to configure Mouse Handler=Basic, Pad1=Connected and Pad2=Null.\n• Marvel Super Hero Squad: Comic Combat  requires to configure Mouse Handler=Basic, Pad1=Null and Pad2=Connected.");
```

Comment (Florin9doi) [reply rpcs3/rpcs3qt/tooltips.h:234]: I added 2 spaces to improve the legibility, Without rich text this was the method I found. 
I found only 3 PS3 games with uDraw support (Wii have more), and the third game works with any of these 2 settings.

**Pull request merged**